app_server.py

```diff
@@ -239,218 +239,108 @@ def heartbeat(self):
-                CoverageRecord.identifier==identifier).filter(
-                    CoverageRecord.data_source_id.in_(
-                        [x.id for x in metadata_sources]))
-        coverage_records = q.all()
-        unaccounted_for = set(metadata_sources)
-        for r in coverage_records:
-            if r.data_source in unaccounted_for:
-                unaccounted_for.remove(r.data_source)
-
-        if unaccounted_for:
-            # At least one metadata lookup has not successfully
-            # completed.
-            names = [x.name for x in unaccounted_for]
-            logging.info(
-                "Cannot build metadata-based OPDS feed for %r: missing coverage records for %s",
-                identifier,
-                ", ".join(names)
-            )
-            unresolved_identifier, is_new = UnresolvedIdentifier.register(
-                self._db, identifier)
-            if is_new:
-                # We just found out about this identifier, or rather,
-                # we just found out that someone expects it to be associated
-                # with a LicensePool.
-                return (201, self.IDENTIFIER_REGISTERED)
-            else:
-                # There is a pending attempt to resolve this identifier.
-                message = (unresolved_identifier.exception 
-                           or self.WORKING_TO_RESOLVE_IDENTIFIER)
-                return (unresolved_identifier.status, message)
-        else:
-            # All metadata lookups have completed. Create that OPDS
-            # entry!
-            entry = identifier.opds_entry()
-
-        if entry is None:
-            # This app can't do lookups on an identifier of this
-            # type, so the best thing to do is to treat this
-            # identifier as a 404 error.
-            return (404, self.UNRECOGNIZED_IDENTIFIER)
-
-        # We made it!
-        return entry
-
-    def work_lookup(self, annotator, route_name='lookup',
-                    require_active_licensepool=True, collection=None):
+    def work_lookup(self, annotator, route_name='lookup', collection=None,
+                    urns=[]):
         """Generate an OPDS feed describing works identified by identifier."""
         urns = flask.request.args.getlist('urn')
 
         this_url = cdn_url_for(route_name, _external=True, urn=urns)
         for urn in urns:
-            self.process_urn(urn, collection=collection)
-
-        # The commit is necessary because we may have registered new
-        # Identifier or UnresolvedIdentifier objects.
-        self._db.commit()
+            self.process_urn(urn, collection)
+        self.post_lookup_hook()
 
         opds_feed = LookupAcquisitionFeed(
             self._db, "Lookup results", this_url, self.works, annotator,
             messages_by_urn=self.messages_by_urn, 
             precomposed_entries=self.precomposed_entries,
-            require_active_licensepool=require_active_licensepool
```

Comment:
Content and circulation always need an active LicensePool for the Work. Metadata always needs an active LicensePool for the Work _except_ when doing a lookup based on ISBN, which we do with a precomposed entry instead of a Work.